### ISSUE TITLE
train model,  kobj_loss is always 0

### ISSUE THREAD

**Issue by Jeremiah0425** [ultralytics/ultralytics#6193]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

![image](https://github.com/ultralytics/ultralytics/assets/51948524/223c7e8e-d6d2-4d47-81b2-fb4068bdfef9)
When I use pose model to train face key points, kobj_loss is always 0

this is args：
task: pose
mode: train
model: model\yolov8n-pose.pt
data: cfg\datasets\wider_face.yaml
epochs: 300
patience: 50
batch: 16
imgsz: 640
save: true
save_period: -1
cache: false
device: 0
workers: 8
project: null
name: null
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: false
cos_lr: false
close_mosaic: 10
resume: false
amp: true
fraction: 1.0
profile: false
overlap_mask: true
mask_ratio: 4
dropout: 0.0
val: true
split: val
save_json: false
save_hybrid: false
conf: null
iou: 0.7
max_det: 300
half: false
dnn: false
plots: true
source: null
show: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
vid_stride: 1
line_width: null
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
boxes: true
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: false
opset: null
workspace: 4
nms: false
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
label_smoothing: 0.0
nbs: 64
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 0.0
translate: 0.1
scale: 0.5
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.5
mosaic: 1.0
mixup: 0.0
copy_paste: 0.0
cfg: null
tracker: botsort.yaml
save_dir: runs\pose\train6
![image](https://github.com/ultralytics/ultralytics/assets/51948524/a0036e85-e6be-40d7-bbcd-55185b1a7135)
this is my data label:
![image](https://github.com/ultralytics/ultralytics/assets/51948524/92ec32b3-6af3-41f5-81de-ff07241a698e)

**Comment by glenn-jocher** [comment]
@Jeremiah0425 hello,

Thank you for reaching out with your question regarding the `kobj_loss` remaining 0 during your training with the pose model. This could be an indication of a couple of potential issues that are worth exploring.

Firstly, please ensure that your dataset labels are correctly formatted and that they include keypoints for the face that the pose model is meant to detect. Incorrectly formatted labels or missing keypoint information could lead to the model not effectively learning to predict keypoints, which might be reflected in the `kobj_loss`.

Secondly, check that the `num_classes` in your dataset configuration aligns with the number of keypoints you’re trying to predict. A mismatch here might mean that keypoint information is being overlooked during the training process.

Also, review your data augmentation settings to make sure they are appropriate for keypoint detection. Overly aggressive augmentations can distort keypoints beyond recognition, preventing effective learning.

Another point to consider is the initialization of the model. If you are using a pretrained model, verify that its pretraining is suitable for keypoint detection tasks.

Lastly, ensure that your model is receiving a mix of positive and negative samples. Only having images with keypoints (or without) could result in a skewed learning that doesn't properly reflect the task complexity.

If everything seems configured correctly, consider running your training with increased verbosity or debugging enabled; this may provide further insights into what's happening during training.

Should the issue persist after checking all of the above, I would recommend examining the loss function implementation to confirm it is behaving as expected. Sometimes, there may be subtle bugs or implementation details that cause unexpected behavior.

Remember, diagnosing such issues can sometimes be a process of elimination. If you continue to face difficulties, please feel free to provide further details so that we can assist you more effectively. I hope this guidance helps you in resolving the issue and moving forward with your training.

Best regards,
Glenn Jocher

**Comment by Jeremiah0425** [comment]
kobj_loss always 0 related to the fact that there are labels <0 in my  labels?
I want to train face detection and key point detection at the same time, but some data only have faces and no key points. Can I use negative numbers instead of key point coordinates?

**Comment by glenn-jocher** [comment]
Hello @Jeremiah0425,

Thank you for posting your issue regarding `kobj_loss` remaining at 0 during your training with YOLOv8 for face detection and keypoint detection tasks.

The presence of labels with negative numbers in your dataset might indeed be contributing to the issue you're experiencing with `kobj_loss`. In keypoint detection tasks, labels are generally expected to have non-negative values that represent pixel coordinates within the image boundaries. Negative numbers may not be interpreted correctly by the model, potentially leading to unexpected behavior such as a loss value of zero.

When you have data that only has faces without keypoints, you typically have two options:
1. You can label the keypoints as occluded or not visible, if your model and labeling format support this concept. This way, the model learns to differentiate between visible and non-visible keypoints.
2. You may use a placeholder value that your model interprets as "keypoint not present" or "ignore." This is different from using negative numbers and would usually be a pre-defined value or method agreed upon within the model's design for representing absent keypoints.

For face detection and keypoint detection on the same model, ensure that you have a consistent labeling schema across your dataset, so that even when keypoints are not available, the model can still learn face detection effectively. It is crucial that the absence of keypoints is handled in a way that the model can recognize and process during training.

In short, avoid using negative numbers for keypoint coordinates, and either find a way to mark keypoints as invisible in a way supported by your model or use a consistent placeholder. Adjust your dataset accordingly, and this should help in resolving the `kobj_loss` staying at zero.

If you need further assistance, consider reviewing the documentation on the correct format for labeling keypoints in datasets intended for use with YOLOv8, or reach out to the community for more detailed guidance on your labeling strategy.

Best regards and happy training!

**Comment by Jeremiah0425** [comment]
Using placeholders, do I need to modify the data preprocessing script in the source code and how to modify it? Thank you!

**Comment by glenn-jocher** [comment]
Hello again @Jeremiah0425,

Using placeholders for keypoints in datasets where some keypoints are absent is indeed a common practice, and it's important that your model's data loading and preprocessing stages can correctly interpret these placeholders.

For YOLOv8, if you choose to use placeholders for missing keypoints, you normally don't need to directly modify the data preprocessing script in the source code as long as the placeholder you use is within the expected range of the input data. Typically, you could use a value that doesn't normally occur as a real keypoint coordinate, such as `0` or some representative value within the image dimensions which is understood to mean "keypoint absent."

Ensure your choice of placeholder does not collide with real, meaningful data. Depending on the model implementation, some models might expect a specific flag or symbol to denote absent keypoints, while others might be configured to ignore keypoints that are outside the image boundaries, for example with coordinates set to `-1`.

If your model requires that absent keypoints are handled in a specific way that is not by default supported by your data format, you may then need to adjust the dataset parsing code to recognize your chosen placeholder and convert it to the format the model expects.

While modifying the source code isn't usually necessary, if you need to customize the behavior of the data loader, you would look for the part of the preprocessing code that parses your dataset annotation files. There, you could add a conditional statement to check for your placeholder values and handle them appropriately.

When making modifications to YOLOv8's source code, ensure you have a good understanding of Python and PyTorch, and always test your changes thoroughly. You could create a simple script to verify that your dataset is being loaded correctly and that the placeholder values are being interpreted as you intended.

Keep in mind that documentation and code comments can be very helpful when navigating the source code and making such adjustments.

Train safe and good luck with your face and keypoint detection tasks!

**Comment by JasseurHadded1** [comment]
Hello @glenn-jocher 

I have the same problem (kobj_loss = 0 during the training),

You said: 

> Secondly, check that the num_classes in your dataset configuration aligns with the number of keypoints you’re trying to predict. A mismatch here might mean that keypoint information is being overlooked during the training process.

Just to be sure, is the num_classes here equal to number of dataset classes (ex: 1 if It's just human pose estimation), or the number of keypoints?

Thanks

**Comment by glenn-jocher** [comment]
@yy-jarvis hello there! 👋

It looks like the issue with `kobj_loss` being 0 might be related to the format of your dataset. YOLOv8 pose model supports both 2D and 3D keypoint annotations. For 2D annotations, you might encounter `kobj_loss=0` if there's a format mismatch. Please ensure your keypoints are correctly annotated and visible in the images. Here's a quick example for 2D keypoints in the expected format:

```yaml
# Example of 2D keypoints annotation format
- image_path.jpg
  keypoints: [x1, y1, visibility1, x2, y2, visibility2, ...]
```

Make sure the `visibility` flag is set correctly for each keypoint (e.g., 0 for not visible/absent, 1 for visible). This should help resolve the issue. Happy coding! 😊

**Comment by al-park** [comment]
@glenn-jocher 

In your example above you say it is for 2D keypoints, and then show the format as (x, y, visibility). Can I check that is correct (not the 3D format)?

On the [Ultralytics website](https://docs.ultralytics.com/tasks/pose/) it says:

> The locations of the keypoints are usually represented as a set of 2D [x, y] or 3D [x, y, visible] coordinates.

Not trying to be a pedant here. I am getting `kobj_loss = 0` and can't figure out why. I ended up on this thread and wanted to double check the format. I am using 2D (x,y) format.

Thanks

**Comment by glenn-jocher** [comment]
Thank you for pointing that out. For 2D keypoints, the format should indeed be [x, y] without the visibility flag. If you're using this format and still seeing `kobj_loss = 0`, please ensure your dataset annotations align with this structure.
